### PR TITLE
[keycloak] Use password file instead of env var for db password

### PR DESCRIPTION
**Pull request by zeph**
any other solution using env variables will cause the bash to interpret the password and possibly fail

### CODE REVIEW

**Comment by zeph** [comment]
@unguiculus @thomasdarimont  anything I can help out with to push this through?

File: charts/keycloak/templates/_helpers.tpl [hunk 1]
```diff
@@ -166,11 +166,16 @@ Create environment variables for database configuration.
   value: {{ .Values.keycloak.persistence.dbName | quote }}
 - name: DB_USER
   value: {{ .Values.keycloak.persistence.dbUser | quote }}
+{{- if .Values.keycloak.persistence.existingSecret}}
+- name: DB_PASSWORD_FILE
+  value: /secrets/persistence
+{{- else }}
```
Comment (unguiculus) [inline charts/keycloak/templates/_helpers.tpl]: I'd suggest we completely move to using password files only. This is already the case for the Keycloak password.

Comment (zeph) [reply charts/keycloak/templates/_helpers.tpl]: I don't know if anybody wants to still use only the env variable, the docker container allows people to do so... I tried to be backward compatible

Comment (unguiculus) [reply charts/keycloak/templates/_helpers.tpl]: I don't think we will have a compatibilty issue here. I think an upgrade should not be an issue.

Comment (zeph) [reply charts/keycloak/templates/_helpers.tpl]: your call... I don't have visibility on all possible hacks done by people on their systems out there

**Comment by zeph** [comment]
@unguiculus what changes shall I address?

**Comment by zeph** [comment]
@unguiculus would u mind solve or answer the CR u opened? 
...then I'll look at these conflicts that just pop up

**Comment by zeph** [comment]
hi @unguiculus , hope u are better... is there any follow up on this issue?

**Comment by unguiculus** [comment]
@zeph Sorry for the delay. Not sure what I need to clarify. As mentioned in another comment, I'd like you to switch to password files completely instead of making this optional.

**Comment by zeph** [comment]
@unguiculus oh, ok... so... shall I modify the code? 
u wanna take the time to? how shall we move on?

**Comment by unguiculus** [comment]
Ah, I thought you could do it, but if you want me to do it I might be able to spend some time on it today.

**Comment by zeph** [comment]
@unguiculus if you don't mind... at least u get the way u have it in mind

**Comment by unguiculus** [comment]
OK, started looking into it.

**Comment by unguiculus** [comment]
@zeph Please have a look and test it.

**Comment by zeph** [comment]
@unguiculus I'm on it right now, thanks for the contribution

**Comment by zeph** [comment]
@unguiculus I believe we should not put the username as part of the Secret, there is no need of encrypting such thing... it is just adding an unnecessary level of complexity

**Comment by unguiculus** [comment]
@zeph The username for the database is already in the secret. There were PRs for this a while ago. I think that's ok. Anyways, this is not related to this PR. However, it's a little inconsistent now because only the DB user is in the secret wheras the Keycloak user is not.

**Comment by zeph** [comment]
@unguiculus sorry, it doesn't make any sense to have the username in secret. 
I reverted your changes as they ended up in my fork (interesting behaviour of github).
...I don't get why you see it inconsistent, Secret is not meant to be used for non-sensitive information. I'm using `SealedSecrets` for that, it would mean I have to have an sealed version of non-required-to-be-encrypted value in each `values` file I have (one per environment)

Would you mind keeping the pull request minimal and just merge it?

**Comment by zeph** [comment]
@thomasdarimont @lenzenowski can you please have a look at
 this pull request? seems we reached a dead end in the discussion

**Comment by zeph** [comment]
@gzur I see the DB_USER is in the Secret as u proposed it... what's the added value?

**Comment by gzur** [comment]
@zeph 
> @gzur I see the DB_USER is in the Secret as u proposed it... what's the added value?

In short, it's so that Keycloak can rely on credentials provisioned dynamically by an operator.

More info:  https://github.com/codecentric/helm-charts/pull/132

**Comment by zeph** [comment]
@gzur do we agree that a username doesn't require to be encrypted, right?

**Comment by gzur** [comment]
> @gzur do we agree that a username doesn't require to be encrypted, right?

What is the value of **not** storing the Username in a Secret?

**Comment by zeph** [comment]
@gzur avoiding the need of having one encrypted value per cluster if using a proper tool like SealedSecrets

KISS (Keep It Simple and Stupid) is the base principle we shall use to work. There is absolutely NO need to put the DB_USER into the Secret element, aside of obfuscating it further and preventing people to use secret management tools

**Comment by zeph** [comment]
@gzur do we agree on the #kiss principle?

**Comment by gzur** [comment]
You seem to be describing a workflow I'm unfamiliar with because decoupling the Username from the Password (and thus having separate delivery mechanisms for each part of the credential pair) doesn't sound very KISS to me, and actually violates The Principle of Least Astonishment, from my point of view. 

Again, I am unfamiliar with the workflow you describe, and I have no insight into how well it gels with the larger Helm ecosystem, but I get the feeling that you are letting some rather specific configuration pain introduced by your upstream toolchain trickle into this downstream repository.

**Comment by zeph** [comment]
@gzur I'm not sure if this TPLA is made up or you are serious. Helm ecosystem allows me to reference the codecentric helm repository in a requirements.yaml file, right? Good, I did that. Sounds pretty Helm ecosystem standard to me. Now, in my helm chart, I do have a template for a SealedSecret resource, I keep the value for the password in its environment yaml file.

Now, encrypting the username seems a practice you alone love. Please point me where anybody else has been doing it, as I'm pretty not aware of such practice, sounds quite over-engineering to me.

**Comment by gzur** [comment]
I have no great love -nor even an opinion - on encrypting usernames since Kubernetes Secrets do not imply encryption in any way.

I implemented https://github.com/codecentric/helm-charts/pull/132 to meet needs that I and others using Vault and other operator based credential provisioning/management systems had.

If my contribution is not to your liking, you are free to make PR to revert it.

Best of luck with your pull request.

**Comment by zeph** [comment]
@gzur I'm trying to understand your reason, before getting into a fight pushing a pull request that you are already bullying... I don't see the point, at all, to encrypt a username. From what you last stated seems you are using Vault also for the username, is such? Are you using it to automatically rotate along with the password?

Before we had a clear interface, an existing Secret resource with a password. Now we have a different interface, with both in... feels to me you are the one that broke the common interface for everyone else

**Comment by unguiculus** [comment]
I'm going to close this PR now. I won't merge it as is. @zeph The way you are acting here doesn't lead anywhere. Please watch the tone in your words and study our code of conduct: https://github.com/codecentric/helm-charts/blob/master/CODE_OF_CONDUCT.md

**Pull request closed without merging**